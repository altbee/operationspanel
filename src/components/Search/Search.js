import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

export const Search = props => {
  const { handleChange, placeholder } = props;

  return (
    <div className="search-component">
      <div className="input-group search">
        <input
          type="text"
          className="form-control input"
          placeholder={placeholder}
          onChange={e => handleChange(e.target.value)}
        />
        <FontAwesomeIcon icon="search" />
      </div>
    </div>
  );
};

Search.propTypes = {
  handleChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string
};

Search.defaultProps = {
  placeholder: "Search Term"
};

export default Search;
