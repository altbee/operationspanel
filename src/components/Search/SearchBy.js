import React, { Component } from "react";
import PropTypes from "prop-types";
import DatePicker from "react-datepicker";
import moment from "moment/moment";

import "react-datepicker/dist/react-datepicker.css";

export default class SearchBy extends Component {
  constructor(props) {
    super(props);

    this.state = {
      opened: false,
      selectedFilter: props.filters[0] || { name: "", type: "text" },
      fromDate: new moment(),
      toDate: new moment(),
      previousSearch: null
    };

    this.searchTimeout = null;
  }

  prepareSearch(data) {
    clearTimeout(this.searchTimeout);
    this.searchTimeout = setTimeout(() => {
      const { onSearch } = this.props;
      const { selectedFilter } = this.state;
      onSearch(data, selectedFilter);
      this.setState({ previousSearch: data });
    }, 500);
  }

  handleTextChange(event) {
    this.prepareSearch(event.target.value.toLowerCase());
  }

  handleDateChange(field, date) {
    this.setState(
      {
        [field]: date
      },
      () => {
        const { fromDate, toDate } = this.state;
        this.prepareSearch({
          fromDate: moment(fromDate).startOf("day"),
          toDate: moment(toDate).endOf("day")
        });
      }
    );
  }

  toggleDropdown() {
    const { opened } = this.state;
    this.setState({ opened: !opened });
  }

  selectFilter(filter) {
    const { onDropdownChange, onSearch } = this.props;
    const { selectedFilter, previousSearch } = this.state;

    let shouldClearPreviousSearch = selectedFilter.type !== filter.type;

    this.setState({
      selectedFilter: filter,
      previousSearch: shouldClearPreviousSearch ? null : previousSearch
    });

    onDropdownChange(filter);

    if (!shouldClearPreviousSearch) {
      onSearch(previousSearch, filter);
    }
  }

  render() {
    const { filters, classes } = this.props;
    const { opened, selectedFilter, fromDate, toDate } = this.state;
    const dropdownClass = opened ? "show" : "";

    return (
      <div className={`search-component ${classes.map(d => d).join(" ")}`}>
        <div className="input-group">
          <input
            type="hidden"
            name="search_param"
            value="all"
            id="search_param"
          />
          {selectedFilter.type === "text" && (
            <input
              type="text"
              className="form-control input has-dropdown"
              placeholder="Search term by..."
              onChange={e => this.handleTextChange(e)}
            />
          )}
          {selectedFilter.type === "date" && [
            <div className="input-group-prepend" key="searchDateLabelFrom">
              <span className="input-group-text">From:</span>
            </div>,
            <DatePicker
              key="searchDatePickerFrom"
              selected={fromDate}
              onChange={date => this.handleDateChange("fromDate", date)}
              dateFormat="L"
              className="form-control"
            />,
            <div className="input-group-prepend" key="searchDateLabelTo">
              <span className="input-group-text">To:</span>
            </div>,
            <DatePicker
              key="searchDatePickerTo"
              selected={toDate}
              onChange={date => this.handleDateChange("toDate", date)}
              dateFormat="L"
              className="form-control"
            />
          ]}
          <div className="input-group-btn search-panel dropdown-wrapper">
            <div
              title={`Search By ${selectedFilter.name}`}
              className="btn-group"
              onClick={this.toggleDropdown.bind(this)}
              role="button"
              tabIndex={-1}
              onKeyPress={() => {}}
            >
              <button
                type="button"
                className="btn btn-danger dropdown-toggle search-by-btn-fw"
                data-toggle="dropdown"
                aria-haspopup="true"
                aria-expanded="false"
              >
                {selectedFilter.name}
              </button>

              <div className={`dropdown-menu ${dropdownClass}`}>
                {filters.map(d => (
                  <button
                    className="dropdown-item"
                    key={d.name}
                    onKeyPress={() => {}}
                    onClick={() => this.selectFilter(d)}
                    type="button"
                  >
                    {d.name}
                  </button>
                ))}
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

SearchBy.propTypes = {
  filters: PropTypes.instanceOf(Array).isRequired,
  classes: PropTypes.instanceOf(Array),
  onSearch: PropTypes.func.isRequired,
  onDropdownChange: PropTypes.func
};

SearchBy.defaultProps = {
  classes: [],
  onDropdownChange: () => {}
};
