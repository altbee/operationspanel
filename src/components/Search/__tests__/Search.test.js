import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Search } from "../Search";

configure({ adapter: new Adapter() });

describe("<Search />", () => {
  const handleChange = value => value;
  const hWord = "Hello World";
  const props = { handleChange, placeholder: hWord };

  it("renders with all props defined", () => {
    const rendered = shallow(<Search {...props} />);
    expect(rendered.find("input").props().placeholder).toBe(hWord);
  });

  it("renders with default props", () => {
    const rendered = shallow(<Search handleChange={handleChange} />);
    expect(rendered.find("input").props().placeholder).toBe("Search Term");
  });

  it("calls the defined handleChange function", () => {
    const onChangeMocked = jest.fn();
    const e = {
      preventDefault() {},
      target: { value: "called" }
    };

    const rendered = shallow(<Search handleChange={onChangeMocked} />);
    rendered.find("input").simulate("change", e);
    expect(onChangeMocked).toBeCalledWith(e.target.value);
  });

  it("matches the snapshot", () => {
    const tree = shallow(<Search {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
