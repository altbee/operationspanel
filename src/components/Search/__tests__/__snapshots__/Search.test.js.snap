// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> matches the snapshot 1`] = `
<div
  className="search-component"
>
  <div
    className="input-group search"
  >
    <input
      className="form-control input"
      onChange={[Function]}
      placeholder="Hello World"
      type="text"
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="search"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
</div>
`;
